File: backend/scanner.py
"""Scan directories for movie files and parse titles from filenames."""

import os
import re
from pathlib import Path
from typing import Optional
from nfo import read_nfo

VIDEO_EXTENSIONS = {
    ".mkv", ".mp4", ".avi", ".m4v", ".wmv", ".flv", ".mov",
    ".mpg", ".mpeg", ".ts", ".m2ts", ".divx", ".ogm", ".webm",
}

SAMPLE_PATTERNS = [re.compile(p, re.IGNORECASE) for p in [r"sample", r"trailer", r"extras?[/\\]", r"featurettes?[/\\]"]]

# Matches multi-part suffixes: CD1, CD2, Part1, Part 2, Disc1, Disk 2, etc.
MULTIPART_RE = re.compile(r'[\s._-]*(cd|part|pt|disc|disk)[\s._-]*([0-9]+)\s*$', re.IGNORECASE)

DVD_FOLDER_NAMES = {"video_ts", "bdmv", "backup", "certificate"}
DVD_EXTENSIONS = {".vob", ".ifo", ".bup", ".vro"}

# --- Filename cleaning (ported from Media Companion) ---

CLEAN_TAGS = [
    "ac3", "aac", "dts", "atmos", "divx", "xvid", "x264", "x265", "h264", "h265",
    "hevc", "dvdrip", "bluray", "blu-ray", "bdrip", "brrip", "bdremux", "remux",
    "dvdscr", "screener", "fullscreen", "widescreen", "telesync", "telecine",
    "hdtv", "webrip", "web-dl", "webdl", "vodrip", "hdrip",
    "480p", "576p", "720p", "1024p", "1080p", "2160p", "4k", "uhd",
    "480", "576", "720", "1024", "1080", "2160",
    "hdr", "hdr10", "10bit", "dv", "remastered", "extended", "unrated",
    "proper", "repack", "internal", "limited",
]

CLEAN_TAGS_SEP_PREFIX = ["scr", "ts", "fs", "ws", "r5"]
CLEAN_TAGS_MULTIWORD = ["special edition", "directors cut", "dir cut", "director's cut"]
CLEAN_MULTIPART = ["part", "pt", "cd", "dvd", "disk", "disc"]
RELEASE_FORMATS = [
    "cam", "telesync", "workprint", "telecine", "pay-per-view rip",
    "screener", "r5", "dvd-rip", "dvd-r", "hdtv", "vodrip",
    "brrip", "bdrip", "bluray", "dvd", "webdl", "webrip",
]

SEP = r"[\-_. ]"


def parse_movie_name(name: str) -> dict:
    """Parse a movie filename into title and year, using Media Companion's algorithm."""
    # Strip extension
    filename = re.sub(r"\.(mkv|mp4|avi|m4v|wmv|flv|mov|mpg|mpeg|ts|m2ts|divx|ogm|webm)$", "", name, flags=re.IGNORECASE)
    # Replace dots and underscores with spaces
    filename = filename.replace(".", " ").replace("_", " ")

    cut = len(filename)

    # 1: Multipart
    m = re.search(rf"(?:{'|'.join(CLEAN_MULTIPART)})(?:{SEP}?)[0-9a-z]", filename, re.IGNORECASE)
    if m and m.start() < cut:
        cut = m.start()

    # 2: DVD5/9
    m = re.search(rf"dvd{SEP}?[59]", filename, re.IGNORECASE)
    if m and m.start() < cut:
        cut = m.start()

    # 3: Sep-prefix tags
    m = re.search(rf"({SEP})(?:{'|'.join(CLEAN_TAGS_SEP_PREFIX)})(?:{SEP}|$)", filename, re.IGNORECASE)
    if m and m.start() < cut:
        cut = m.start()

    # 4: Standard tags
    m = re.search(rf"(?:{SEP}|\[|\()?(?:{'|'.join(CLEAN_TAGS)})(?:{SEP}|\]|\)|$)", filename, re.IGNORECASE)
    if m and m.start() < cut:
        cut = m.start()

    # 5: Multi-word tags
    for mw in CLEAN_TAGS_MULTIWORD:
        pattern = mw.replace(" ", SEP)
        m = re.search(pattern, filename, re.IGNORECASE)
        if m and m.start() < cut:
            cut = m.start()

    # 6: Release formats
    m = re.search(rf"(?:{SEP}|\[|\()(?:{'|'.join(RELEASE_FORMATS)})(?:{SEP}|\]|\)|$)", filename, re.IGNORECASE)
    if m and m.start() < cut:
        cut = m.start()

    # 7: Year detection (Media Companion priority)
    year = None
    year_cut = None

    # a) Year in parentheses
    ym = re.search(r"\(((?:19|20)\d{2})\)", filename)
    if ym:
        year = int(ym.group(1))
        year_cut = ym.start()

    # b) Year in brackets
    if year is None:
        ym = re.search(r"\[((?:19|20)\d{2})\]", filename)
        if ym:
            year = int(ym.group(1))
            year_cut = ym.start()

    # c) Year with separator before AND after (won't match at position 0)
    if year is None:
        ym = re.search(r"[\-_. \[(]((?:19|20)\d{2})[\-_. \])]", filename)
        if ym:
            year = int(ym.group(1))
            year_cut = ym.start()

    if year_cut is not None and year_cut < cut:
        cut = year_cut

    # Truncate
    if 0 < cut < len(filename):
        filename = filename[:cut]

    # Clean up
    filename = re.sub(r"[\-_. ]+$", "", filename).strip()
    filename = re.sub(r"\s+", " ", filename).strip()

    return {"title": filename or name, "year": year}


def _strip_multipart(name: str) -> tuple[str, int | None]:
    """Return (base_title, part_number) stripping CD1/CD2/Part1 etc."""
    m = MULTIPART_RE.search(name)
    if m:
        return name[:m.start()].strip(), int(m.group(2))
    return name, None


def scan_directories(directories: list[str]) -> list[dict]:
    """Scan directories for movie files, grouping multi-part movies."""
    raw = []
    counter = 0

    for directory in directories:
        if not os.path.isdir(directory):
            continue
        for root, dirs, files in os.walk(directory):
            # Skip DVD/BD folder structures in-place (prunes os.walk)
            dirs[:] = [d for d in dirs if d.lower() not in DVD_FOLDER_NAMES]
            # Also skip if current folder IS a DVD structure
            if os.path.basename(root).lower() in DVD_FOLDER_NAMES:
                continue
            for fname in sorted(files):  # sorted so CD1 comes before CD2
                filepath = os.path.join(root, fname)
                ext = os.path.splitext(fname)[1].lower()
                if ext in DVD_EXTENSIONS:
                    continue
                if ext not in VIDEO_EXTENSIONS:
                    continue
                if any(p.search(filepath) for p in SAMPLE_PATTERNS):
                    continue
                try:
                    size = os.path.getsize(filepath)
                except OSError:
                    continue
                if size < 50 * 1024 * 1024:
                    continue

                counter += 1
                name_no_ext = os.path.splitext(fname)[0]
                base_name, part_num = _strip_multipart(name_no_ext)
                parsed = parse_movie_name(base_name if part_num is not None else name_no_ext)
                folder_path = root
                folder_name = os.path.basename(root)

                # Check for existing companion files
                nfo_path = _find_companion(filepath, folder_path, [".nfo"])
                poster_path = _find_image(filepath, folder_path, "poster")
                fanart_path = _find_image(filepath, folder_path, "fanart")

                movie_data = None
                matched = False
                tmdb_id = None
                imdb_id = None
                if nfo_path:
                    movie_data = read_nfo(nfo_path)
                    if movie_data:
                        matched = True
                        tmdb_id = movie_data.get("tmdbId")
                        imdb_id = movie_data.get("imdbId")

                raw.append({
                    "id": f"movie-{counter}",
                    "filePath": filepath,
                    "fileName": fname,
                    "folderPath": folder_path,
                    "folderName": folder_name,
                    "parsedTitle": parsed["title"],
                    "parsedYear": parsed["year"],
                    "hasNfo": nfo_path is not None,
                    "hasPoster": poster_path is not None,
                    "hasFanart": fanart_path is not None,
                    "nfoPath": nfo_path,
                    "posterPath": poster_path,
                    "fanartPath": fanart_path,
                    "matched": matched,
                    "ignored": False,
                    "tmdbId": tmdb_id,
                    "imdbId": imdb_id,
                    "movieData": movie_data,
                    "_partNum": part_num,
                })

    # --- Group multi-part movies ---
    movies = []
    grouped: dict[tuple, dict] = {}  # (folderPath, parsedTitle, parsedYear) -> primary entry

    for m in raw:
        part_num = m.pop("_partNum", None)
        if part_num is not None:
            key = (m["folderPath"], m["parsedTitle"], m["parsedYear"])
            if key in grouped:
                # Add as additional part to existing primary entry
                primary = grouped[key]
                primary["parts"].append({"partNum": part_num, "filePath": m["filePath"], "fileName": m["fileName"]})
                # Prefer part1 companion files; fill in if primary is missing them
                if not primary["hasNfo"] and m["hasNfo"]:
                    primary["hasNfo"] = m["hasNfo"]
                    primary["nfoPath"] = m["nfoPath"]
                    primary["matched"] = m["matched"]
                    primary["movieData"] = m["movieData"]
                    primary["tmdbId"] = m["tmdbId"]
                    primary["imdbId"] = m["imdbId"]
            else:
                m["parts"] = [{"partNum": part_num, "filePath": m["filePath"], "fileName": m["fileName"]}]
                grouped[key] = m
                movies.append(m)
        else:
            m["parts"] = []
            movies.append(m)

    return movies


def _find_companion(video_path: str, folder_path: str, extensions: list[str]) -> Optional[str]:
    name_no_ext = os.path.splitext(os.path.basename(video_path))[0]
    for ext in extensions:
        by_name = os.path.join(folder_path, name_no_ext + ext)
        if os.path.exists(by_name):
            return by_name
        generic = os.path.join(folder_path, "movie" + ext)
        if os.path.exists(generic):
            return generic
    return None


def _find_image(video_path: str, folder_path: str, img_type: str) -> Optional[str]:
    name_no_ext = os.path.splitext(os.path.basename(video_path))[0]
    for ext in [".jpg", ".jpeg", ".png", ".webp"]:
        by_name = os.path.join(folder_path, f"{name_no_ext}-{img_type}{ext}")
        if os.path.exists(by_name):
            return by_name
        generic = os.path.join(folder_path, f"{img_type}{ext}")
        if os.path.exists(generic):
            return generic
    return None

File: src/bun/scanner.ts
import { readdirSync, statSync, existsSync } from "fs";
import { join, extname, basename, dirname, parse } from "path";
import type { ScannedMovie } from "./types";

const VIDEO_EXTENSIONS = new Set([
  ".mkv", ".mp4", ".avi", ".m4v", ".wmv", ".flv", ".mov",
  ".mpg", ".mpeg", ".ts", ".m2ts", ".divx", ".ogm", ".webm",
]);

const SAMPLE_PATTERNS = [/sample/i, /trailer/i, /extras?[/\\]/i, /featurettes?[/\\]/i];

export function scanDirectories(directories: string[]): ScannedMovie[] {
  const movies: ScannedMovie[] = [];
  let idCounter = 0;

  for (const dir of directories) {
    if (!existsSync(dir)) continue;
    const found = findMovieFiles(dir);
    for (const filePath of found) {
      idCounter++;
      const movie = analyzeMovieFile(filePath, `movie-${idCounter}`);
      movies.push(movie);
    }
  }

  return movies;
}

function findMovieFiles(directory: string): string[] {
  const files: string[] = [];

  function walk(dir: string) {
    try {
      const entries = readdirSync(dir);
      for (const entry of entries) {
        const fullPath = join(dir, entry);
        try {
          const stat = statSync(fullPath);
          if (stat.isDirectory()) {
            walk(fullPath);
          } else if (stat.isFile()) {
            const ext = extname(entry).toLowerCase();
            if (VIDEO_EXTENSIONS.has(ext) && !isSampleFile(fullPath) && stat.size > 50 * 1024 * 1024) {
              files.push(fullPath);
            }
          }
        } catch {
          // skip inaccessible files
        }
      }
    } catch {
      // skip inaccessible directories
    }
  }

  walk(directory);
  return files;
}

function isSampleFile(filePath: string): boolean {
  return SAMPLE_PATTERNS.some((pattern) => pattern.test(filePath));
}

function analyzeMovieFile(filePath: string, id: string): ScannedMovie {
  const fileName = basename(filePath);
  const folderPath = dirname(filePath);
  const folderName = basename(folderPath);
  const { title, year } = parseMovieName(parse(fileName).name);

  const nfoPath = findCompanionFile(filePath, [".nfo"]);
  const posterPath = findImageFile(filePath, folderPath, "poster");
  const fanartPath = findImageFile(filePath, folderPath, "fanart");

  return {
    id,
    filePath,
    fileName,
    folderPath,
    folderName,
    parsedTitle: title,
    parsedYear: year,
    hasNfo: nfoPath !== null,
    hasPoster: posterPath !== null,
    hasFanart: fanartPath !== null,
    nfoPath,
    posterPath,
    fanartPath,
    matched: false,
    tmdbId: null,
    imdbId: null,
    movieData: null,
  };
}

// --- Filename cleaning ported from Media Companion's Utilities.CleanFileName ---

const CLEAN_SEPARATORS = "-_. ";
const SEP_REGEX_CLASS = `[\\-_. ]`;

// Tags that don't need a separator before them (case-insensitive)
const CLEAN_TAGS = [
  "ac3", "aac", "dts", "atmos", "divx", "xvid", "x264", "x265", "h264", "h265",
  "hevc", "dvdrip", "bluray", "blu-ray", "bdrip", "brrip", "bdremux", "remux",
  "dvdscr", "screener", "fullscreen", "widescreen", "telesync", "telecine",
  "hdtv", "webrip", "web-dl", "webdl", "vodrip", "hdrip",
  "480p", "576p", "720p", "1024p", "1080p", "2160p", "4k", "uhd",
  "480", "576", "720", "1024", "1080", "2160",
  "hdr", "hdr10", "10bit", "dv", "remastered", "extended", "unrated",
  "proper", "repack", "internal", "limited",
];

// Tags that must have a separator character before them
const CLEAN_TAGS_SEP_PREFIX = ["scr", "ts", "fs", "ws", "r5"];

// Multi-word tags (spaces become separator wildcards)
const CLEAN_TAGS_MULTIWORD = ["special edition", "directors cut", "dir cut", "director's cut"];

// Multipart indicators
const CLEAN_MULTIPART = ["part", "pt", "cd", "dvd", "disk", "disc"];

// Release formats
const RELEASE_FORMATS = [
  "cam", "telesync", "workprint", "telecine", "pay-per-view rip",
  "screener", "r5", "dvd-rip", "dvd-r", "hdtv", "vodrip",
  "brrip", "bdrip", "bluray", "dvd", "webdl", "webrip",
];

function parseMovieName(name: string): { title: string; year: number | null } {
  // Strip file extension if present
  let filename = name.replace(/\.(mkv|mp4|avi|m4v|wmv|flv|mov|mpg|mpeg|ts|m2ts|divx|ogm|webm)$/i, "");

  // Replace dots and underscores with spaces (like MC does first)
  filename = filename.replace(/\./g, " ").replace(/_/g, " ");

  let cutPosition = filename.length;

  // 1: Check multipart tags (e.g. "cd1", "part 1", "disc1")
  const multipartPattern = new RegExp(
    `((?:${CLEAN_MULTIPART.join("|")})(?:${SEP_REGEX_CLASS}?)[0-9a-z])`, "i"
  );
  let m = filename.match(multipartPattern);
  if (m && m.index !== undefined && m.index < cutPosition) {
    cutPosition = m.index;
  }

  // 2: Check dvd5 / dvd9 tags
  m = filename.match(/dvd[\-_. ]?[59]/i);
  if (m && m.index !== undefined && m.index < cutPosition) {
    cutPosition = m.index;
  }

  // 3: Check tags that must have a separator prefix
  const sepPrefixPattern = new RegExp(
    `(${SEP_REGEX_CLASS}(?:${CLEAN_TAGS_SEP_PREFIX.join("|")}))(?:${SEP_REGEX_CLASS}|$)`, "i"
  );
  m = filename.match(sepPrefixPattern);
  if (m && m.index !== undefined && m.index < cutPosition) {
    cutPosition = m.index;
  }

  // 4: Check standard tags (no separator required)
  const tagsPattern = new RegExp(
    `(?:${SEP_REGEX_CLASS}|\\[|\\()?(?:${CLEAN_TAGS.join("|")})(?:${SEP_REGEX_CLASS}|\\]|\\)|$)`, "i"
  );
  m = filename.match(tagsPattern);
  if (m && m.index !== undefined && m.index < cutPosition) {
    cutPosition = m.index;
  }

  // 5: Check multi-word tags (with separator wildcards)
  for (const mw of CLEAN_TAGS_MULTIWORD) {
    const mwPattern = new RegExp(mw.replace(/ /g, `${SEP_REGEX_CLASS}`), "i");
    m = filename.match(mwPattern);
    if (m && m.index !== undefined && m.index < cutPosition) {
      cutPosition = m.index;
    }
  }

  // 6: Check release formats
  const relPattern = new RegExp(
    `(?:${SEP_REGEX_CLASS}|\\[|\\()(?:${RELEASE_FORMATS.join("|")})(?:${SEP_REGEX_CLASS}|\\]|\\)|$)`, "i"
  );
  m = filename.match(relPattern);
  if (m && m.index !== undefined && m.index < cutPosition) {
    cutPosition = m.index;
  }

  // 7: Extract year using Media Companion's priority approach:
  //    a) Year in parentheses: (2024)
  //    b) Year in brackets: [2024]
  //    c) Year with separator before AND after (won't match at position 0)
  //    This prevents titles like "1978" or "2001" from being eaten as the year.
  let year: number | null = null;
  let yearCut: number | null = null;

  // a) Check for year in parentheses first — most reliable
  let ym = filename.match(/\(((?:19|20)\d{2})\)/);
  if (ym && ym.index !== undefined) {
    year = parseInt(ym[1], 10);
    yearCut = ym.index;
  }

  // b) Check for year in square brackets
  if (year === null) {
    ym = filename.match(/\[((?:19|20)\d{2})\]/);
    if (ym && ym.index !== undefined) {
      year = parseInt(ym[1], 10);
      yearCut = ym.index;
    }
  }

  // c) Check for year with a separator before AND after (not at position 0)
  if (year === null) {
    ym = filename.match(/[\-_. \[(]((?:19|20)\d{2})[\-_. \])]/);
    if (ym && ym.index !== undefined) {
      year = parseInt(ym[1], 10);
      yearCut = ym.index;
    }
  }

  if (yearCut !== null && yearCut < cutPosition) {
    cutPosition = yearCut;
  }

  // Truncate at the earliest tag position
  if (cutPosition > 0 && cutPosition < filename.length) {
    filename = filename.substring(0, cutPosition);
  }

  // Clean trailing separators and extra whitespace
  filename = filename.replace(/[\-_. ]+$/, "").replace(/\s+/g, " ").trim();

  // Replace remaining hyphens surrounded by spaces for readability
  filename = filename.replace(/ - /g, " - ");

  return { title: filename || name, year };
}

function findCompanionFile(videoPath: string, extensions: string[]): string | null {
  const dir = dirname(videoPath);
  const nameNoExt = parse(videoPath).name;

  for (const ext of extensions) {
    // Check movie-name.nfo
    const byName = join(dir, nameNoExt + ext);
    if (existsSync(byName)) return byName;

    // Check movie.nfo (folder-based)
    const generic = join(dir, "movie" + ext);
    if (existsSync(generic)) return generic;
  }
  return null;
}

function findImageFile(videoPath: string, folderPath: string, type: "poster" | "fanart"): string | null {
  const nameNoExt = parse(videoPath).name;
  const imageExts = [".jpg", ".jpeg", ".png", ".webp"];

  for (const ext of imageExts) {
    // filename-poster.jpg
    const byName = join(folderPath, `${nameNoExt}-${type}${ext}`);
    if (existsSync(byName)) return byName;

    // poster.jpg (folder-based)
    const generic = join(folderPath, `${type}${ext}`);
    if (existsSync(generic)) return generic;
  }
  return null;
}
